workerextension.go

```diff
@@ -27,154 +24,73 @@ import (
-	// The returned request will be passed to the worker script.
-	GetRequest() *WorkerRequest
-	// SendRequest sends a request to the worker script. The callback function of frankenphp_handle_request() will be called.
-	SendRequest(r *WorkerRequest)
-}
-
-// EXPERIMENTAL: WorkerRequest represents a request to pass to a worker script.
-type WorkerRequest struct {
-	// Request is an optional HTTP request for your worker script to handle
-	Request *http.Request
-	// Response is an optional response writer that provides the output of the provided request, it must not be nil to access the request body
-	Response http.ResponseWriter
-	// CallbackParameters is an optional field that will be converted in PHP types or left as-is if it's an unsafe.Pointer and passed as parameter to the PHP callback
-	CallbackParameters any
-	// AfterFunc is an optional function that will be called after the request is processed with the original value, the return of the PHP callback, converted in Go types, is passed as parameter
-	AfterFunc func(callbackReturn any)
+type Worker struct {
+	name     string
+	fileName string
+	num      int
+	options  []WorkerOption
 }
 
 var extensionWorkers = make(map[string]Worker)
-var extensionWorkersMutex sync.Mutex
 
 // EXPERIMENTAL: RegisterWorker registers a custom worker script.
 func RegisterWorker(worker Worker) {
-	extensionWorkersMutex.Lock()
-	defer extensionWorkersMutex.Unlock()
-
-	extensionWorkers[worker.Name()] = worker
+	extensionWorkers[worker.name] = worker
 }
 
-// startWorker creates a pipe from a worker to the main worker.
-func startWorker(w *worker, extensionWorker Worker, thread *phpThread) {
-	for {
-		rq := extensionWorker.GetRequest()
-
-		var fc *frankenPHPContext
-		if rq.Request == nil {
-			fc = newFrankenPHPContext()
-			fc.logger = logger
-		} else {
-			fr, err := NewRequestWithContext(rq.Request, WithOriginalRequest(rq.Request))
-			if err != nil {
-				logger.LogAttrs(context.Background(), slog.LevelError, "error creating request for external worker", slog.String("worker", w.name), slog.Int("thread", thread.threadIndex), slog.Any("error", err))
-				continue
-			}
-
-			var ok bool
-			if fc, ok = fromContext(fr.Context()); !ok {
-				continue
-			}
-		}
-
-		fc.worker = w
-
-		fc.responseWriter = rq.Response
-		fc.handlerParameters = rq.CallbackParameters
-
-		// Queue the request and wait for completion if Done channel was provided
-		logger.LogAttrs(context.Background(), slog.LevelInfo, "queue the external worker request", slog.String("worker", w.name), slog.Int("thread", thread.threadIndex))
-
-		w.requestChan <- fc
-		if rq.AfterFunc != nil {
-			go func() {
-				<-fc.done
-
-				if rq.AfterFunc != nil {
-					rq.AfterFunc(fc.handlerReturn)
-				}
-			}()
-		}
-	}
-}
+// EXPERIMENTAL: SendRequest sends an HTTP request to the worker and writes the response to the provided ResponseWriter.
+func (w Worker) SendRequest(rw http.ResponseWriter, r *http.Request) error {
```

Comment:
Do we mean to make the receiver here a non-pointer? I'm not 100% sure, but this would mean (potentially) a copy needs to be made every time a message is sent.

Reply:
Anyway, that's a micro-optimisation (though this is on a hot-path). I think I have a good idea here, I'll create a PR to this PR -- but I do need to leave for a date in the next hour. I think this is fine for now, so it isn't a blocker from merging.